### PR TITLE
Added monaco and YAML support to devfile editor

### PR DESCRIPTION
**Pull request by JPinkney**
### What does this PR do?
Iit adds YAML support for the Che devfile editor. Some big changes introduced are:
	1. Introducing a new dashboard API point for accessing the devfile schema
	2. Deprecating codemirror in favour of Monaco in other places that the editor is used

2 is done because I don't think it makes sense to run multiple editors in the dashboard + have all their dependencies loaded + maintain two editors

Screenshot from 2019-11-01 15-24-28

##### Things left
- [X] Changing the monaco editors height and width (currently set at 600px by 1000px for testing purposes)
- [x] Modifying the workspace devfile editor controller
- [x] Fix the yarn.lock
- [x] Set tab width to 2

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13939

#### Release Notes

Introduces a Che devfile editor with autocompletion, validation, and hover

### CODE REVIEW

**Comment by benoitf** [comment]
ci-build

File: dashboard/package.json [hunk 1]
```diff
@@ -84,6 +84,7 @@
     "protractor-jasmine2-screenshot-reporter": "^0.3.3",
     "proxy-middleware": "^0.15.0",
     "raw-loader": "^0.5.1",
+    "regenerator-runtime": "^0.13.3",
```
Comment (benoitf) [inline dashboard/package.json]: do we have CQs for these new dependencies ?

Comment (JPinkney) [reply dashboard/package.json]: Not yet, I'm not a Che commiter so someone else has to submit the CQ's

**Comment by benoitf** [comment]
@JPinkney FYI build is failing with error message
```
 15:48:50 [ERROR] Failed to execute goal com.mycila:license-maven-plugin:3.0:check (check-headers) on project che-dashboard-war: Some files do not have the expected license header -> [Help 1]
15:48:50 [ERROR] 
```

**Comment by l0rd** [comment]
@JPinkney would you be able to address https://github.com/eclipse/che/issues/15175 in this PR? Do you see any inconvenience?

**Comment by tsmaeder** [comment]
> 
> 
> @JPinkney would you be able to address #15175 in this PR? Do you see any inconvenience?

I do: we want to get this merged and have no work planned in this sprint beyond that.

**Review by akurinnoy — COMMENTED** [review]
Great work!

**Comment by akurinnoy** [comment]
@JPinkney during manual testing I faced places where editor doesn't work:
- creating factory from template
- factory configuration on a factory details page
- workspace configuration while exporting a workspace

dashboard-monaco-editor

**Comment by tsmaeder** [comment]
@akurinnoy thanks for testing! To make it simpler to debug, we could use some info:
1. When you mean "doesn't work", you mean the editor is so small you cannot see anything or is there any other misbehaviour you can see?
2. What browser/system do you use?
3. Is there anythign interesting in the browser log?

**Comment by benoitf** [comment]
@tsmaeder AFAIK there is animated gif that is showing what is not working (editor not displayed)

**Comment by benoitf** [comment]
ci-build

**Comment by akurinnoy** [comment]
@tsmaeder 

macOs Mojave v10.14.6
Google Chrome 78.0.3904.97 (64bit)

1. while creating factory from template editor is in DOM but it doesn't work completely, no errors in browser console.
2. on factory details page editor has very small hight, I can see some content, no errors in browser
3. while exporting a workspace there is no editor at all and some logs in console:

<details>
<summary>error</summary>

```
Error: ModelService: Cannot add model because it already exists!
    at ModelServiceImpl._createModelData (modelServiceImpl.js:335)
    at ModelServiceImpl.createModel (modelServiceImpl.js:345)
    at doCreateModel (standaloneEditor.js:85)
    at Object.createModel (standaloneEditor.js:102)
    at UiMonaco.link (monaco.directive.ts:71)
    at angular.js:1393
    at invokeLinkFn (angular.js:10682)
    at nodeLinkFn (angular.js:10071)
    at compositeLinkFn (angular.js:9311)
    at nodeLinkFn (angular.js:10065) "<div flex="" class="workspace-editor ng-pristine ng-untouched ng-valid" ui-monaco="exportWorkspaceDialogController.editorOptions" ng-model="exportWorkspaceDialogController.exportConfigContent">"
(anonymous) @ angular.js:14961
(anonymous) @ angular.js:11281
invokeLinkFn @ angular.js:10684
nodeLinkFn @ angular.js:10071
compositeLinkFn @ angular.js:9311
nodeLinkFn @ angular.js:10065
compositeLinkFn @ angular.js:9311
publicLinkFn @ angular.js:9176
link @ angular-material.js:24192
(anonymous) @ angular.js:1393
(anonymous) @ angular.js:10676
invokeLinkFn @ angular.js:10682
nodeLinkFn @ angular.js:10071
compositeLinkFn @ angular.js:9311
publicLinkFn @ angular.js:9176
lazyCompilation @ angular.js:9567
boundTranscludeFn @ angular.js:9354
controllersBoundTransclude @ angular.js:10121
ngIfWatchAction @ angular.js:28487
$digest @ angular.js:18543
$apply @ angular.js:18903
done @ angular.js:12775
completeRequest @ angular.js:13032
requestLoaded @ angular.js:12937
load (async)
(anonymous) @ angular.js:12920
sendReq @ angular.js:12720
serverRequest @ angular.js:12461
processQueue @ angular.js:17330
(anonymous) @ angular.js:17378
$digest @ angular.js:18515
$apply @ angular.js:18903
done @ angular.js:12775
completeRequest @ angular.js:13032
requestLoaded @ angular.js:12937
load (async)
(anonymous) @ angular.js:12920
sendReq @ angular.js:12720
serverRequest @ angular.js:12461
processQueue @ angular.js:17330
(anonymous) @ angular.js:17378
$digest @ angular.js:18515
$apply @ angular.js:18903
(anonymous) @ angular.js:27961
dispatch @ jquery.js:5237
elemData.handle @ jquery.js:5044
```
</details>

**Comment by tsmaeder** [comment]
thx, @akurinnoy

File: dashboard/package.json (hunk 1 above)
Comment (tsmaeder) [reply dashboard/package.json]: @JPinkney did we ever file that CQ? What packages do we need to file a cq for?

File: dashboard/src/components/attribute/monaco/monaco.directive.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      monacoOptions.onLoad(editor);
+    }
+  }
+
+  private configOptionsWatcher(editor: any, monacoDefaults: string[], uiMonacoAttr: string, scope: ng.IScope): void {
+    if (!uiMonacoAttr) {
+      return;
+    }
+
+    scope.$watch(uiMonacoAttr, (newValues: Object, oldValue: Object) => {
```
[the remark this answers is not in the document]
Comment (JPinkney) [reply dashboard/src/components/attribute/monaco/monaco.directive.ts:114]: Its the same way in master: https://github.com/eclipse/che/blob/master/dashboard/src/components/attribute/codemirror/codemirror.directive.ts#L94 (the new monaco directive was modified from the codemirror one)

File: dashboard/package.json (hunk 1 above)
Comment (JPinkney) [reply dashboard/package.json]: I don't think we filed it. I'm not sure what determines which packages we need to file cq's for but this is a list of the new dependencies with their licenses: 

regenerator-runtime - MIT
glob - ISC
monaco-editor-core - MIT
monaco-languageclient - MIT
monaco-languages - MIT
umd-compat-loader - Apache-2.0
vscode-json-languageservice - MIT
vscode-languageserver-types - MIT
yaml-language-server - MIT

File: dashboard/src/components/attribute/monaco/monaco.directive.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/components/attribute/monaco/monaco.directive.ts]: It was just a wrong type after copy/paste when declared a constant. I will add a fix.

**Comment by olexii4** [comment]
I have prepared a docker image from this branch and push it to upstream docker.io/olexii4dockerid/che-server:nightly

I deploy CHE with chectl and add the custom image
--cheimage=docker.io/olexii4dockerid/che-server:nightly

All good.

**Review by olexii4 — APPROVED** [review]

**Comment by benoitf** [comment]
ci-build

**Comment by AndrienkoAleksandr** [comment]
@JPinkney since you reworked dashboard to don't use CodeMirror, you can remote dependency from package.json to clean:
https://github.com/eclipse/che/pull/15067/files#diff-8a93884670d48666f7da0af337381d92R127

File: dashboard/src/components/api/che-devfile.factory.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  /**
+   * Retrieve the current JSON Schema of the devfile
+   */
+  fetchDevfileSchema(): ng.IPromise<any> {
+    if (this.devfileJSONSchemaPromise) {
+      return this.devfileJSONSchemaPromise;
+    }
+
+    let promise = this.$http.get('/api/devfile/');
```
Comment (AndrienkoAleksandr) [inline dashboard/src/components/api/che-devfile.factory.ts:40]: Does this request cover https?

File: dashboard/src/components/attribute/monaco/monaco.directive.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2015-2018 Red Hat, Inc.
```
Comment (AndrienkoAleksandr) [inline dashboard/src/components/attribute/monaco/monaco.directive.ts:2]: Minor, but 2019

File: dashboard/src/components/api/che-devfile.factory.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2015-2018 Red Hat, Inc.
```
Comment (AndrienkoAleksandr) [inline dashboard/src/components/api/che-devfile.factory.ts:2]: 2019 and take a look another files, please.

File: dashboard/src/components/attribute/monaco/monaco.directive.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+const UI_MONACO_CONFIG = {
+  wordWrap: 'on',
+  lineNumbers: 'on',
+  matchBrackets: true,
+  autoClosingBrackets: 'always',
+  readOnly: false
+};
+
+/**
+ * Binds a Monaco widget to a textarea element.
```
Comment (AndrienkoAleksandr) [inline dashboard/src/components/attribute/monaco/monaco.directive.ts]: Monaco editor was embedded to the div, so maybe `textarea` -> `div`

File: dashboard/src/components/attribute/monaco/monaco.directive.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      $scope.$eval(uiMonaco),
+      {
+        model: Monaco.editor.createModel('',  'yaml'),
+        automaticLayout: true
+      });
+
+    const editor = Monaco.editor.create(element, monacoOptions);
+    (window as any).MonacoEditor = editor;
+
+    // this sets the default height of the editor to be consistent with the default of codemirror
```
Comment (olexii4) [inline dashboard/src/components/attribute/monaco/monaco.directive.ts]: Minor one  codemirror -> monaco

File: dashboard/src/components/widget/editor/che-editor.controller.ts [hunk 8]
```diff
@@ -155,10 +126,67 @@ export class CheEditorController {
...
+
+    const yamlService = (window as any).yamlService;
+    const pendingValidationRequests = new Map();
+
+    const getModel = () => monaco.editor.getModels()[0];
+
+    const cleanPendingValidation = (document) => {
+        const request = pendingValidationRequests.get(document.uri);
+        if (request !== undefined) {
+        clearTimeout(request);
```
Comment (AndrienkoAleksandr) [inline dashboard/src/components/widget/editor/che-editor.controller.ts:158]: Format code indentation here, please.

File: dashboard/src/app/workspaces/workspace-details/environments/environments.controller.ts [hunk 9]
```diff
@@ -83,13 +84,14 @@ export class WorkspaceEnvironmentsController {
...
+      wordWrap: 'on',
+      lineNumbers: 'on',
       readOnly: true,
-      gutters: [],
+      lineDecorationsWidth: 0,
+      lineNumbersMinChars: 0,
       onLoad: (editor: any) => {
         $timeout(() => {
-          editor.refresh();
+          editor.render(true);
```
Comment (olexii4) [inline dashboard/src/app/workspaces/workspace-details/environments/environments.controller.ts]: We added this timeout depends on the CodeMirror bug. I think we can remove it at all (in all parts of the code).
WDYT?

**Comment by olexii4** [comment]
We should clean up code from the old CodeMirror specific styles in *.styl files like this one https://github.com/JPinkney/che/blob/monaco/dashboard/src/app/workspaces/workspaces.styl#L5

File: dashboard/src/components/attribute/monaco/monaco.directive.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (olexii4) [reply dashboard/src/components/attribute/monaco/monaco.directive.ts:114]: No. It isn`t a mistake, just look like this. I look into the code - the string is the right type.  It is an attribute and returns the name of the string variable of the parent controller.  For example 'cheEditorController.editorOptions'.  
It becomes an object after $scope.$eval

**Comment by dmytro-ndp** [comment]
crw-ci-test

File: dashboard/src/app/workspaces/workspace-details/environments/environments.controller.ts (hunk 9 above)
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/environments/environments.controller.ts]: Let it be a separate issue.

**Review by AndrienkoAleksandr — APPROVED** [review]

**Comment by olexii4** [comment]
@JPinkney Could you fix dependency versions
```
    "glob": "^7.1.5",
    "monaco-editor-core": "^0.14.6",
    "monaco-languageclient": "^0.9.0",
    "monaco-languages": "^1.8.0",
    "umd-compat-loader": "^2.1.1",
    "vscode-languageserver-types": "^3.10.0",
```
to 
```
    "glob": "7.1.5",
    "monaco-editor-core": "0.14.6",
    "monaco-languageclient": "0.9.0",
    "monaco-languages": "1.8.0",
    "umd-compat-loader": "2.1.1",
    "vscode-languageserver-types": "3.10.0",
```
?

**Comment by olexii4** [comment]
@JPinkney 

I created CQs:
  https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21300

**Comment by olexii4** [comment]
@JPinkney  And fix this:

``` "regenerator-runtime": "^0.13.3" ``` -> ``` "regenerator-runtime": "0.13.3" ```

File: dashboard/package.json (hunk 1 above)
Comment (olexii4) [reply dashboard/package.json]: created

**Review by akurinnoy — APPROVED** [review]

**Review by benoitf — APPROVED** [review]

**Comment by dmytro-ndp** [comment]
@JPinkney: [Happy path tests failed](https://github.com/eclipse/che/pull/15067#issuecomment-566523151) because there are outdated in the PR. Please, take latest changes from Eclipse Che **master** branch.

**Comment by olexii4** [comment]
crw-ci-test

**Comment by olexii4** [comment]
crw-ci-test

**Comment by olexii4** [comment]
ci-build

**Comment by olexii4** [comment]
ci-test

**Comment by olexii4** [comment]
ci-test

**Comment by olexii4** [comment]
ci-build

**Pull request merged**